rootfs/etc/nginx/lua/balancer.lua

```diff
@@ -0,0 +1,112 @@
+local json = require "json"
+local b = require "ngx.balancer"
+
+local function compare(t, fn)
+    if #t == 0 then return nil, nil end
+    local key, value = 1, t[1]
+    for i = 2, #t do
+        if fn(value, t[i]) then
+            key, value = i, t[i]
+        end
+    end
+    return key, value
+end
+
+local http_host = ngx.var.http_host
+local remote_addr = ngx.var.remote_addr
+local proxy_upstream_name = ngx.var.proxy_upstream_name
+local shared_memory = ngx.shared.shared_memory;
+local cfgs_json = shared_memory:get("CFGS")
+if cfgs_json == nil then
+    ngx.status = 503
+    ngx.exit(ngx.status)
+end
+local cfgs = json.decode(cfgs_json)
+local loadbalancealgorithm = cfgs.cfg.loadbalancealgorithm
+local backend = cfgs.backends[proxy_upstream_name]
+if backend == nil then
+    ngx.status = 503
+    ngx.exit(ngx.status)
+end
+local selected_endpoint
+if backend.sessionaffinity.affinitytype == "cookie" then
+    local backend_cookie_name = backend.sessionaffinity.cookiesessionaffinity.name
+    local backend_cookie_hash = backend.sessionaffinity.cookiesessionaffinity.hash
+    local var_name = "cookie_SESSION_" .. backend_cookie_name .. backend_cookie_hash
+    
+    local cookie_value = ngx.var[var_name]
+
+    if cookie_value == nil then
+        local endpoints_roundrobin = ngx.shared.endpoints_roundrobin
+        local ep_index = endpoints_roundrobin:get(http_host)
```

Comment:
Should load balancing states not be scoped to backend/upstream instead? If I read this correctly you are load balancing per http_host rather than upstream. Which for example means given two domains and 5 endpoints, if I send a request to first domain and another request to the second domain they will both be proxied to the same endpoint.